app/src/org/commcare/dalvik/activities/HomeActivityUIController.java

```diff
@@ -0,0 +1,404 @@
+        };
+    }
+
+    private View.OnClickListener getLogoutButtonListener() {
+        return new View.OnClickListener() {
+            public void onClick(View v) {
+                CommCareApplication._().closeUserSession();
+                activity.returnToLogin();
+            }
+        };
+    }
+
+    // endregion
+
+
+    // region - text setters for all buttons
+
+    private void setSyncButtonText(String syncTextKey) {
+        if (syncTextKey == null) {
+            syncTextKey = activity.isDemoUser() ? "home.sync.demo" : "home.sync";
+        }
+        if (numUnsentForms > 0) {
+            Spannable syncIndicator = (activity.localize("home.sync.indicator",
+                    new String[]{String.valueOf(numUnsentForms), Localization.get(syncTextKey)}));
+            syncButton.setNotificationText(syncIndicator);
+            adapter.notifyDataSetChanged();
+        } else {
+            syncButton.setText(activity.localize(syncTextKey));
+        }
+    }
+
+    private void setIncompleteFormsText() {
+        if (numIncompleteForms > 0) {
+            Spannable incompleteIndicator = (activity.localize("home.forms.incomplete.indicator",
+                    new String[]{String.valueOf(numIncompleteForms), Localization.get("home.forms.incomplete")}));
+            if (viewIncompleteFormsButton != null) {
+                viewIncompleteFormsButton.setText(incompleteIndicator);
+            }
+        } else {
+            if (viewIncompleteFormsButton != null) {
+                viewIncompleteFormsButton.setText(activity.localize("home.forms.incomplete"));
+            }
+        }
+    }
+
+    // endregion
+
+
+    protected void refreshView() {
+        refreshVersionText();
+        refreshButtonTextSources();
+        refreshHomeAndLogoutButtons();
+
+        refreshDataFromSyncDetails();
+        setIncompleteFormsText();
+        refreshSyncButton();
+        showSyncMessage();
+
+        setButtonVisibilities();
+        activity.updateCommCareBanner();
+        adapter.notifyDataSetChanged();
+    }
+
+
+    // region - all helper methods used by refreshView()
+
+    private void refreshVersionText() {
+        TextView version = (TextView)activity.findViewById(R.id.str_version);
+        if (version == null) {
+            return;
+        }
+        version.setText(CommCareApplication._().getCurrentVersionString());
+    }
+
+    private void refreshButtonTextSources() {
+        if (activity.isDemoUser()) {
+            syncKey = "home.sync.demo";
+            lastMessageKey = "home.sync.message.last";
+            homeMessageKey = "home.start.demo";
+            logoutMessageKey = "home.logout.demo";
+        } else {
+            syncKey = "home.sync";
+            lastMessageKey = "home.sync.message.last";
+            homeMessageKey = "home.start";
+            logoutMessageKey = "home.logout";
+        }
+    }
+
+    private void refreshHomeAndLogoutButtons() {
+        if (startButton != null) {
```

Comment:
I feel like we should be able to take out a lot of these nullity checks. Let's keep this in mind for a future refactor.